cpp/src/stream_compaction/unique_count.cu

```diff
@@ -0,0 +1,138 @@
+
+namespace cudf {
+namespace detail {
+namespace {
+/**
+ * @brief A functor to be used along with device type_dispatcher to check if
+ * the row `index` of `column_device_view` is `NaN`.
+ */
+struct check_nan {
+  // Check if it's `NaN` for floating point type columns
+  template <typename T, std::enable_if_t<std::is_floating_point_v<T>>* = nullptr>
+  __device__ inline bool operator()(column_device_view const& input, size_type index)
+  {
+    return std::isnan(input.data<T>()[index]);
+  }
+  // Non-floating point type columns can never have `NaN`, so it will always return false.
+  template <typename T, std::enable_if_t<not std::is_floating_point_v<T>>* = nullptr>
+  __device__ inline bool operator()(column_device_view const&, size_type)
+  {
+    return false;
+  }
+};
+}  // namespace
+
+cudf::size_type unique_count(table_view const& keys,
+                             null_equality nulls_equal,
+                             rmm::cuda_stream_view stream)
+{
+  auto table_ptr = cudf::table_device_view::create(keys, stream);
+  row_equality_comparator comp(
+    nullate::DYNAMIC{cudf::has_nulls(keys)}, *table_ptr, *table_ptr, nulls_equal);
+  return thrust::count_if(
+    rmm::exec_policy(stream),
+    thrust::counting_iterator<cudf::size_type>(0),
+    thrust::counting_iterator<cudf::size_type>(keys.num_rows()),
+    [comp] __device__(cudf::size_type i) { return (i == 0 or not comp(i, i - 1)); });
+}
+
+cudf::size_type unique_count(column_view const& input,
+                             null_policy null_handling,
+                             nan_policy nan_handling,
+                             rmm::cuda_stream_view stream)
+{
+  auto const num_rows = input.size();
+
+  if (num_rows == 0 or num_rows == input.null_count()) { return 0; }
+
+  auto const count_nulls      = null_handling == null_policy::INCLUDE;
+  auto const nan_is_null      = nan_handling == nan_policy::NAN_IS_NULL;
+  auto const should_check_nan = cudf::is_floating_point(input.type());
+  auto input_device_view      = cudf::column_device_view::create(input, stream);
+  auto device_view            = *input_device_view;
+  auto input_table_view       = table_view{{input}};
+  auto table_ptr              = cudf::table_device_view::create(input_table_view, stream);
+  row_equality_comparator comp(nullate::DYNAMIC{cudf::has_nulls(input_table_view)},
+                               *table_ptr,
+                               *table_ptr,
+                               null_equality::EQUAL);
+
+  return thrust::count_if(
+    rmm::exec_policy(stream),
+    thrust::counting_iterator<cudf::size_type>(0),
+    thrust::counting_iterator<cudf::size_type>(num_rows),
+    [count_nulls, nan_is_null, should_check_nan, device_view, comp] __device__(cudf::size_type i) {
+      auto const is_null = device_view.is_null(i);
+      auto const is_nan  = nan_is_null and should_check_nan and
+                          cudf::type_dispatcher(device_view.type(), check_nan{}, device_view, i);
```

Comment:
There are two `type_dispatcher()` calls here from within device code, essentially on the same `type`. I wonder if there might be value in dispatching from host, and doing the `count_if()` in the dispatch endpoint?

Reply:
Another consideration is whether to materialize an `is_nan` vector, and look it up in the `count_if` loop.
Is it preferable to incur the cost in computation rather than in temporary memory allocations? (Asking for my own education. :])

Reply:
I've tested materializing `is_nan` vector also, the downside is that will loop the whole input twice. And IIRC, there are no visible performance differences between the current two-type dispatcher method and your propositions.

Reply:
@mythrocks The purpose of `should_check_nan` is to eliminate the device-side type dispatch for non-floating types. We only pay the cost of the device-side type dispatch for floating types that need to be checked for `NaN`.

Reply:
> no visible performance differences between the current two-type dispatcher method and your propositions

Thanks for considering it, @PointKernel. I'll remember this for the next time I have to make a similar choice.